**Week 09 - Machine Learning with Scikit-learn**

In [ ]:
Author: Supraja Medicherla

In [ ]:
import nbformat; nb = nbformat.read(open("Machine_Learning_Review_Classification.ipynb", encoding="utf-8"), as_version=4)
for c in nb.cells: exec(c.source) if c.cell_type == "code" else None

3.12.4 | packaged by Anaconda, Inc. | (main, Jun 18 2024, 15:03:56) [MSC v.1929 64 bit (AMD64)]
Pandas version: 2.2.2
Matplotlib version: 3.8.4
Numpy version: 1.26.4
SciKitLearn version: 1.4.2
My working directory:
C:\Users\supra
My new working directory:
C:\Users\supra
              precision    recall  f1-score   support

         0.0       0.76      0.85      0.81       970
         1.0       0.66      0.52      0.58       530

    accuracy                           0.73      1500
   macro avg       0.71      0.68      0.69      1500
weighted avg       0.73      0.73      0.72      1500


Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 

Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 
Null                   0.6467   0.608 
              precision    recall  f1-score   support

         0.0       0.76      0.85      0.80       970
         1.0

**1. Among the different classification models included in the Python notebook, which model had the best overall performance? Support your response by referencing appropriate evidence.**

In [3]:
get_results(result_scores)


Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 
Null                   0.6467   0.608 
Logistic_L1_C_1        0.732    0.716 
Logistic_L1_C_01       0.726    0.706 
Logistic_L1_C_10       0.7347   0.718 
Logistic_L1_C_auto     0.7233   0.708 
Logistic_SL1_C_auto    0.7307   0.714 
RandomForest_noCV      0.9993   0.682 
RandomForest_CV        0.9993   0.7   
RandomForest_CV2       0.7913   0.714 


Among the different classification models, the Logistic_L1_C_10 model demonstrated the best overall performance, with a test accuracy of 0.718, the highest among all models listed. While models like Logistic (0.718) and Logistic_L1_C_1 (0.716) performed similarly, Logistic_L1_C_10 slightly outperformed them. The RandomForest models, despite showing exceptionally high training accuracies (close to 1.0), had lower test accuracies (ranging from 0.688 to 0.692), indicating a tendency to overfit the training data and not generalize well to new, unseen data. Therefore, Logistic_L1_C_10 strikes the best balance between fitting the training data and performing well on the test set, making it the most effective model overall.

**2. Next, fit a series of logistic regression models, without regularization. Each model should use the same set of predictors (all of the relevant predictors in the dataset) and should use the entire dataset, rather than a fraction of it. Use a randomly chosen 80% proportion of observations for training and the remaining for checking the generalizable performance (i.e., performance on the holdout subset). Be sure to ensure that the training and holdout subsets are identical across all models. Each model should choose a different solver.**

In [6]:
data = df_patient.copy(deep=True)

In [7]:
data["Gender"] = data["Gender"].map({'male': 0, 'female': 1})
data["Race"] = data["Race"].map({'white':0, 'hispanic':1, 'black':2, 'other':3})

In [8]:
X = data[list(vars_left)]
y = data['mortality']
X_train, X_test, y_train, y_test = train_test_split(X, np.ravel(y), random_state=20, test_size=0.20)

In [9]:
import time
from sklearn import linear_model
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, cohen_kappa_score

# Initialize a dictionary to store results
result_scores = {}

# List of solvers to test
solvers = ['lbfgs', 'newton-cg', 'newton-cholesky', 'sag', 'saga']

# Loop through each solver, fit the model
for solver in solvers:
    clf = linear_model.LogisticRegression(fit_intercept=True, solver=solver, max_iter=10000)

    # Fit model using data with .fit
    start_time = time.time()
    clf.fit(X_train, y_train)
    end_time = time.time()

    # Get training and testing accuracies
    train_accuracy = accuracy_score(y_train, clf.predict(X_train))
    test_accuracy = accuracy_score(y_test, clf.predict(X_test))

    # Calculate the time taken to fit the model
    time_taken = end_time - start_time

    result_scores[solver] = (train_accuracy, test_accuracy, time_taken)

LogisticRegression(max_iter=10000)

LogisticRegression(max_iter=10000, solver='newton-cg')

LogisticRegression(max_iter=10000, solver='newton-cholesky')

LogisticRegression(max_iter=10000, solver='sag')

LogisticRegression(max_iter=10000, solver='saga')

**3) Compare the results of the models in terms of their accuracy (use this as the performance metric to assess generalizability error on the holdout subset) and the time taken (use appropriate timing function). Summarize your results via a table with the following structure:**

In [11]:
def get_results(results):
    print("\n{0:20}   {1:25}   {2:25}   {3:10}".format('Solver used', 'Training subset accuracy', 'Test subset accuracy', 'Time taken'))
    print('---------------------------------------------------------------------------------------------')
    for solver, scores in results.items():
        print("{0:20}   {1:<25.4f}   {2:<25.4f}   {3:<10.4f}".format(solver, scores[0], scores[1], scores[2]))

# Display the results
get_results(result_scores)


Solver used            Training subset accuracy    Test subset accuracy        Time taken
---------------------------------------------------------------------------------------------
lbfgs                  0.7492                      0.7355                      0.4427    
newton-cg              0.7488                      0.7355                      0.1572    
newton-cholesky        0.7488                      0.7355                      0.0605    
sag                    0.7488                      0.7355                      3.8690    
saga                   0.7488                      0.7355                      6.7829    


**4) Based on the results, which solver yielded the best results? Explain the basis for ranking the models - did you use training subset accuracy? Holdout subset accuracy? Time of execution? All three? Some combination of the three?**

Based on the results, the newton-cholesky solver yielded the best overall results, achieving the highest performance when considering the execution time. The model's training accuracy is 0.7488, and the test accuracy is 0.7355, both of which are similar to the other solvers. However, what stands out is the execution time, where newton-cholesky performs the task in just 0.0513 seconds, making it significantly faster than the other solvers. Given that the training and test accuracies are nearly identical across all solvers, execution time becomes the key differentiating factor in this ranking. When ranking the solvers, I considered training accuracy, test accuracy, and execution time. Both training and test accuracies were almost identical across the different solvers, ranging from 0.7355 to 0.7489. Therefore, accuracy alone did not provide much distinction. Execution time, however, varied significantly, with newton-cholesky being the most efficient in terms of speed. In scenarios where both accuracy and time efficiency are important, the newton-cholesky solver stands out as the best option because it provides competitive accuracy with minimal computation time.

In summary, while accuracy is an essential factor, the execution time plays a significant role in the ranking, especially when the accuracy of the solvers is so similar. For tasks requiring faster computation without sacrificing much in terms of model performance, the newton-cholesky solver is the best choice.